src/xenia/kernel/xsocket.cc

```diff
@@ -257,5 +257,25 @@ bool XSocket::QueuePacket(uint32_t src_ip, uint16_t src_port,
   return true;
 }
 
+int XSocket::GetSockName(uint8_t* buf, int* buf_len) {
+  struct sockaddr sa = {};
+
+  int ret = getsockname(native_handle_, &sa, (socklen_t*)buf_len);
+
+  if (!ret) {
+    std::memcpy(buf, &sa, *buf_len);
+  }
+  return ret;
+}
+
+const uint32_t XSocket::GetLastWSAError() const {
```

Comment:
I'm not sure if this platform implementation detail — the fact that the host sockets expose a thread-local variable storing the last error — should be revealed to the platform-independent `XSocket` interface. Though we don't have something like `xsocket_posix.cc`/`xsocket_win.cc`, and this file is heavily specific to BSD sockets in its current form which do rely on `errno`/`WSAGetLastError`, but still, it's not ruled out that some future implementation may have error codes reported immediately from the functions as return values or output arguments rather than via a TLS variable, for instance. We already have functions returning `X_STATUS` here — they can be changed to returning 0 on success or the extended error code on error (assuming, however, that the host doesn't for some reason set the last error to 0 on failure in certain cases — if that's the case though, we need to check what the guest expects, and either split the return value into two variables or replace such 0 with a true error code) so there's better isolation of methods in this interface without global state where it's not clearly needed.

Note that we should never ever go in the opposite direction, however (proxying the host `WSAGetLastError` directly to the guest instead of maintaining our own TLS variable on the guest side), as we may in theory use host sockets in implementations of other kernel functions (such as HTTP, device connection).